### ISSUE TITLE
Services allow duplicate `port`

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#47221]
It is possible to express a Service with multiple ports blocks with the same number.  This is not very useful and may cause trouble for implementations of Services.

Curiously, if I `kubectl edit` this and touch that field, patch merges them because the mergeKey is `port`.  @mengqiy it did the right thing and surprised me :)

```
apiVersion: v1
kind: Service
metadata:
  name: hostnames-2
spec:
  ports:
  - name: a
    port: 80
    targetPort: 9376
  - name: b
    port: 80
    targetPort: 9377
  selector:
    run: hostnames
```

**Comment by liggitt** [comment]
> if I kubectl edit this and touch that field, patch merges them because the mergeKey is port

not quite the right thing... if you add distinct protocols, it'll still merge them :-/

this should be valid, right?

```
  ports:
  - name: a
    port: 80
    protocol: UDP
    targetPort: 9376
  - name: b
    port: 80
    protocol: UDP
    targetPort: 9377
```

**Comment by thockin** [comment]
oh, excellent point.  More concretely
https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/kubedns-svc.yaml.base#L31

**Comment by thockin** [comment]
I will open a different issue.

**Comment by mengqiy** [comment]
It will try to apply the patch to the first matching item based on `mergeKey`.
@thockin I bet you must be editing the first item, so it looks working. But if you edit the second item, you will find it doesn't work.
```
  ports:
  - name: a
    port: 80
    targetPort: 9376
  - name: b
    port: 80
    targetPort: 9321
```

**Comment by mengqiy** [comment]
We have a PR to support multifields merge key for new APIs https://github.com/kubernetes/kubernetes/pull/46161 and a proposal to fix the existing APIs https://github.com/kubernetes/community/pull/476.
I really hope we can fix this in 1.8.

**Comment by xiangpengzhao** [comment]
>It is possible to express a Service with multiple ports blocks with the same number. This is not very useful and may cause trouble for implementations of Services.

For this point, per @liggitt comment, should we at least add some validation to verify if "port + protocol" is valid? i.e. "same ports with different protocols" is valid, but "same ports with same protocol" is invalid.

**Comment by liggitt** [comment]
> should we at least add some validation to verify if "port + protocol" is valid? i.e. "same ports with different protocols" is valid, but "same ports with same protocol" is invalid.

I'm not sure. I don't know enough about service implementations to know what kind of problems it can cause, but it is removing flexibility from the service -> container abstraction (and we have all the standard questions around disallowing previously allowed things in the API - will it break updates, will it break status reporting, etc)

**Comment by xiangpengzhao** [comment]
I haven't investigated what it will break yet. But what are users expecting when they configure ports as below and what should we do for such case?

```
  ports:
  - name: a
    port: 80
    protocol: TCP
    targetPort: 9376
  - name: b
    port: 80
    protocol: TCP
    targetPort: 9377
```

**Comment by liggitt** [comment]
ah, I misread... I thought this was routing two service ports to the same container port

**Comment by mengqiy** [comment]
cc: @pwittrock

**Comment by xiangpengzhao** [comment]
This is fixed by #47336. PATCH issue is tracked in #47249. I'm going to close this one.

/close
